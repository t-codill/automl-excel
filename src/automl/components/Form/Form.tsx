import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";
import { FormContext } from "./FormContext";
import { FormDataType } from "./FormDataType";

export interface IFormProps<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
    onUpdated?(key: keyof TData, value: TData[keyof TData]): void;
}
export interface IFormStates<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
}

export class Form<TData extends { [key in keyof TData]: FormDataType }> extends React.Component<IFormProps<TData>, IFormStates<TData>> {
    private readonly _inputs: Set<FormBaseInput<TData, keyof TData, TData[keyof TData]>> = new Set();
    public render(): React.ReactNode {
        return (
            <FormContext.Provider value={{
                mount: this._mount,
                unmount: this._unmount,
                onUpdated: this.props.onUpdated
            }}>
                <form
                    onSubmit={this._onSubmit}
                    noValidate={true}
                    autoComplete="off">
                    {this.props.children}
                </form>
            </FormContext.Provider>
        );

    }
    private readonly _mount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.add(input);
        return;
    }
    private readonly _unmount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.delete(input);
        return;
    }
    private readonly _onSubmit = (event: React.FormEvent<HTMLElement>): void => {
        event.preventDefault();
        const val: TData = Object.create(null);
        let valid = true;
        this._inputs.forEach((input) => {
            val[input.props.field] = input.getValue();
            valid = !input.validate() && valid;
        });
        if (valid) {
            this.props.onSubmit(val);
        }
    }
}
