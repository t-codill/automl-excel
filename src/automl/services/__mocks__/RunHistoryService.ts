import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { classificationSuccessRun } from "../../childRun/__data__/classificationSuccessRun";
import { BasicTypes } from "../../common/BasicTypes";
import { IDictionary } from "../../common/IDictionary";
import { parentSuccessRun } from "../../parentRun/__data__/parentSuccessRun";
import { IArtifact } from "../ArtifactService";

export type IRunDtoWithExperimentName = RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto & { experimentName: string };
export type RunMetricType = BasicTypes | BasicTypes[] | IDictionary<BasicTypes[]>;

export class RunHistoryService {
    public getArtifactBackedMetricNames(): Array<[string, IArtifact]> {
        const artifact: IArtifact = {
            origin: "",
            container: "",
            path: ""
        };
        return [["artifact", artifact]];
    }
    public mergeMetrics(): IDictionary<RunMetricType> {
        if (classificationSuccessRun.runMetrics) {
            return classificationSuccessRun.runMetrics;
        }
        return {};
    }

    public async getExperiment(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto | undefined> {
        return undefined;
    }

    public async getRun(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined> {
        return classificationSuccessRun.run;
    }

    public async getRunMetrics(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[] | undefined> {
        return [classificationSuccessRun.runMetrics];
    }

    public async getChildRuns(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[] | undefined> {
        return undefined;
    }

    public async getChildRunMetrics(): Promise<Array<IDictionary<RunMetricType>> | undefined | undefined> {
        return undefined;
    }

    public async listExperiments(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto[] | undefined> {
        return [
            {
                createdUtc: new Date(2019, 1, 1, 1, 1, 1, 1),
                description: "Description 1",
                experimentId: "Experiment Id 1",
                name: "Experiment Name 1"
            },
            {
                createdUtc: new Date(2019, 2, 2, 2, 2, 2, 2),
                description: "Description 2",
                experimentId: "Experiment Id 2",
                name: "Experiment Name 2"
            },
            {}
        ];
    }

    public async getRunList(): Promise<IRunDtoWithExperimentName[] | undefined> {
        return [
            {
                experimentName: "experiment1",
                runId: "run1"
            },
            {
                experimentName: "experiment2",
                runId: "run2"
            }
        ];
    }

    public async updateTag(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public async updateTags(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
