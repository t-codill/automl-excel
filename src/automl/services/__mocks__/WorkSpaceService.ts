import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { IVMSizeResult } from "../WorkSpaceService";

export class WorkSpaceService {

    private readonly sampleWorkSpace: AzureMachineLearningWorkspacesModels.Workspace = {
        id: "Test_ID",
        name: "Test_WorkSpace",
        location: "eastus",
        discoveryUrl: "https://dummyurl.dummy",
    };

    private readonly sampleCompute: AzureMachineLearningWorkspacesModels.ComputeResource = {
        id: "Test_ID",
        name: "Test_Compute",
        location: "eastus",
    };

    private readonly sampleVMSize: AzureMachineLearningWorkspacesModels.VirtualMachineSize = {
        name: "Standard_DS1_v2",
        family: "standardDSv2Family",
        vCPUs: 1,
        osVhdSizeMB: 1047552,
        maxResourceVolumeMB: 7168,
        memoryGB: 3.5,
        lowPriorityCapable: true,
        premiumIO: true
    };

    public async listWorkspaces(): Promise<AzureMachineLearningWorkspacesModels.Workspace[] | undefined> {
        return [this.sampleWorkSpace];
    }

    public async tryGetWorkspace(): Promise<AzureMachineLearningWorkspacesModels.Workspace | null | undefined> {
        return this.sampleWorkSpace;
    }

    public async getWorkspace(): Promise<AzureMachineLearningWorkspacesModels.Workspace | undefined> {
        return this.sampleWorkSpace;
    }

    public async tryGetCompute(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource | null | undefined> {
        return this.sampleCompute;
    }

    public async listComputes(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined> {
        return [this.sampleCompute];
    }

    public async createCompute(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource | undefined> {
        return this.sampleCompute;
    }

    public async listVmSizes(): Promise<IVMSizeResult | undefined> {
        return { ...[this.sampleVMSize], amlCompute: [this.sampleVMSize] };
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
