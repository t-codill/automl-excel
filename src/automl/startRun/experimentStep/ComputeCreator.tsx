import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { DefaultButton, Panel, PanelType, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { Accordion } from "../../components/Accordion/Accordion";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { Validators } from "../../components/Form/Validators";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ICreateComputeParams } from "./ICreateComputeParams";
import { MaxNodeInput } from "./MaxNodeInput";
import { MinNodeInput } from "./MinNodeInput";
import { VmSizeSelector } from "./VmSizeSelector";

export interface IComputeCreatorProps {
    vmSizes: IVMSizeResult | undefined;
    onComputeCreated(compute: AzureMachineLearningWorkspacesModels.ComputeResource): void;
    onCancel(): void;
}

export interface IComputeCreatorState {
    creating: boolean;
    minNodeCount: string;
}

export class ComputeCreator extends BaseComponent<IComputeCreatorProps, IComputeCreatorState,
    {
        workSpaceService: WorkSpaceService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService
    };

    protected readonly getData = undefined;

    constructor(props: IComputeCreatorProps) {
        super(props);
        this.state = {
            creating: false,
            minNodeCount: "0"
        };
    }

    public render(): React.ReactNode {
        return <>
            {!this.state.creating &&
                <Panel
                    isOpen={true}
                    type={PanelType.medium}
                    onDismiss={this.cancel}
                    isHiddenOnDismiss={true}
                    headerText="Create a New Compute"
                    closeButtonAriaLabel="Close"
                >
                    <Form onSubmit={this.createCompute} onUpdated={this.onFormUpdated}>
                        <FormTextInput<ICreateComputeParams, "name">
                            field="name"
                            required={true}
                            label="Compute name"
                            placeholder="Compute name"
                            ariaLabel="Please enter compute name here"
                            validators={[
                                Validators.required("Compute name is required"),
                                Validators.regex(/^[a-zA-Z][a-zA-Z0-9-]{1,15}$/,
                                    "Compute name should start with a letter, be between 2 and 16 character, and only include letters (a-zA-Z), numbers (0-9) and '-'")]}
                        />
                        <VmSizeSelector vmSizes={this.props.vmSizes} />
                        <Accordion
                            exclusive={true}
                            items={[
                                {
                                    title: "Additional Settings",
                                    collapsed: true,
                                    disabled: false,
                                    element: <>
                                        <MinNodeInput />
                                        <MaxNodeInput minNodeCount={this.state.minNodeCount} />
                                    </>
                                }
                            ]} />
                        <div className="form-footer">
                            <DefaultButton text="Cancel" onClick={this.cancel} />
                            <PrimaryButton text="Create" type="submit" />
                        </div>
                    </Form>
                </Panel>
            }
            {this.state.creating &&
                <PopupProgressIndicator
                    title="Creating a new compute..."
                    description="Creating a new compute. This may take a couple minutes."
                />}
        </>;
    }

    private readonly onFormUpdated = (key: keyof ICreateComputeParams, value: ICreateComputeParams[keyof ICreateComputeParams]) => {
        if (!value) {
            return;
        }
        if (key === "minNodeCount") {
            this.setState({ minNodeCount: value });
        }
    }

    private readonly cancel = () => {
        this.props.onCancel();
    }

    private readonly createCompute = async (data: ICreateComputeParams) => {
        if (!data
            || !data.name
            || !data.vmSize
            || !data.minNodeCount
            || !data.maxNodeCount) {
            return;
        }

        this.setState({
            creating: true
        });
        this.logUserAction("CreateCompute", { data: JSON.stringify(data) });
        const compute = await this.services.workSpaceService.createCompute(
            data.name,
            data.vmSize,
            parseInt(data.minNodeCount, 10),
            parseInt(data.maxNodeCount, 10)
        );

        this.setState({ creating: false });

        if (!compute) {
            return;
        }

        this.props.onComputeCreated(compute);
    }
}
