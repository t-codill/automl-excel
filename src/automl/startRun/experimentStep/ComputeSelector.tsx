import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { reduce } from "lodash";
import { Icon, IDropdownOption, Link } from "office-ui-fabric-react";
import * as React from "react";
import { computeUtils } from "../../common/utils/computeUtils";
import { FormDropdown } from "../../components/Form/FormDropdown";
import { Validators } from "../../components/Form/Validators";
import { LabelWithCallout } from "../../components/Label/LabelWithCallout";
import { IExperimentStepCallback } from "./ExperimentStep";

interface IComputeSelectorProps {
    computes: AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined;
    selectedComputeId: string | undefined;
    readonly: boolean;
    refreshing: boolean;
    onEditClick(): void;
}

export class ComputeSelector extends React.PureComponent<IComputeSelectorProps> {

    public render(): React.ReactNode {
        const dropdownId = "experimentStepComputeSelector";
        return <>
            <LabelWithCallout required={true} htmlFor={dropdownId} labelText="Select a training compute">
                A cloud compute resource is required to run the experiment in Azure.
                You can use your existing Azure ML compute resource, or create a new one.
                Note that compute resource has its own costs, refer to the &nbsp;
                <Link className="ms-CalloutExample-link" href="https://azure.microsoft.com/en-us/pricing/details/machine-learning-service/" target="_blank">
                    pricing details <Icon iconName="NavigateExternalInline" />
                </Link> for more information.
                Computes marked "profiling enabled" support additional data profiling capabilities.</LabelWithCallout>
            <FormDropdown<IExperimentStepCallback, "computeId">
                field="computeId"
                id={dropdownId}
                required={true}
                placeholder="Select a Compute..."
                options={this.loadOptions()}
                buttonIconProps={
                    {
                        iconName: this.getIcon()
                    }
                }
                onClick={this.props.readonly ? this.props.onEditClick : undefined}
                validators={[
                    Validators.required("Compute is required")
                ]}
            />
        </>;
    }
    private readonly getIcon = () => {
        if (this.props.refreshing) {
            return "ProgressRingDots";
        }
        if (this.props.readonly) {
            return "Edit";
        }
        return "ChevronDown";
    }

    private readonly loadOptions = () => {
        return reduce(this.props.computes,
            (options, compute) => this.computeOptionsReducer(options, compute),
            [] as IDropdownOption[]);
    }

    private readonly computeOptionsReducer = (options: IDropdownOption[], compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        const computeUnion = compute.properties;
        if (computeUnion) {
            switch (computeUnion.computeType) {
                case "AmlCompute": {
                    options.push(this.getComputeOptions(compute, computeUtils.isComputeProfilingEnabled(compute)));
                    break;
                }
                case "VirtualMachine": {
                    options.push(this.getComputeOptions(compute, true));
                    break;
                }
                default:
            }
        }
        return options;
    }

    private readonly getComputeOptions = (compute: AzureMachineLearningWorkspacesModels.ComputeResource, profilingEnabled: boolean) => {
        const provisioning = this.getProvisioningInfo(compute);
        const profilingText = this.getProfilingText(profilingEnabled);
        const option: IDropdownOption = {
            key: compute.id || "",
            text: `${compute.name}${provisioning.message}${profilingText}`,
            disabled: !provisioning.isSuccessful
        };
        return option;
    }

    private readonly getProfilingText = (profilingEnabled: boolean) => {
        return profilingEnabled ? " (profiling enabled)" : "";
    }

    private readonly getProvisioningInfo = (compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        const isComputeUsable = computeUtils.isComputeUsable(compute);
        const isProvisioningSuccessful = compute.properties && compute.properties.provisioningState === "Succeeded";
        let message = "";
        if (!isProvisioningSuccessful) {
            message = ` (${(compute.properties && compute.properties.provisioningState) ? compute.properties.provisioningState : "Unknown"})`;
        } else if (!isComputeUsable) {
            message = " (Unusable)";
        }
        return { message, isSuccessful: isProvisioningSuccessful && isComputeUsable };
    }
}
