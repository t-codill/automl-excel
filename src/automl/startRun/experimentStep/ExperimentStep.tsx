import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { map, uniq } from "lodash";
import { DefaultButton, Dialog, DialogFooter, DialogType, Icon, Link, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { RunHistoryService } from "../../services/RunHistoryService";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ComputeCreator } from "./ComputeCreator";
import { ComputeSelector } from "./ComputeSelector";
import { ExperimentNameInput } from "./ExperimentNameInput";

export interface IExperimentStepProps {
    experimentName: string | undefined;
    compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    readOnly: boolean;
    onEditClick(): void;
    onCancel(): void;
    onNext(
        experimentName: string | undefined,
        compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined
    ): void;
}
export interface IExperimentStepCallback {
    experimentName: string | undefined;
    computeId: string | undefined;
}

export interface IExperimentStepState {
    experimentNames: string[] | undefined;
    computes: AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined;
    selectedCompute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    showComputeCreator: boolean;
    vmSizes: IVMSizeResult | undefined;
    showCancelDialog: boolean;
    computeRefreshing: boolean;
}

export class ExperimentStep
    extends BaseComponent<IExperimentStepProps, IExperimentStepState, {
        workSpaceService: WorkSpaceService;
        runHistoryService: RunHistoryService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService,
        runHistoryService: RunHistoryService
    };

    constructor(props: IExperimentStepProps) {
        super(props);
        this.state = {
            experimentNames: undefined,
            computes: undefined,
            selectedCompute: props.compute,
            showComputeCreator: false,
            vmSizes: undefined,
            showCancelDialog: false,
            computeRefreshing: false
        };
    }

    public render(): React.ReactNode {
        return <>
            <Form
                onSubmit={this.onSubmit}>
                <div className="start-run-step-content">
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ExperimentNameInput
                            experimentName={this.props.experimentName}
                            readonly={this.props.readOnly}
                            onEditClick={this.props.onEditClick}
                            experimentNames={this.state.experimentNames}
                        />
                    </div>
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ComputeSelector
                            computes={this.state.computes}
                            selectedComputeId={this.state.selectedCompute && this.state.selectedCompute.id}
                            readonly={this.props.readOnly}
                            refreshing={this.state.computeRefreshing}
                            onEditClick={this.props.onEditClick}
                        />
                        {!this.props.readOnly &&
                            <Link
                                id="linkCreateCompute"
                                onClick={this.onCreateComputeClick}>
                                <Icon iconName="Devices3" style={{ marginRight: 5 }} />
                                <span>Create a new compute</span>
                            </Link>
                        }
                        {!this.props.readOnly &&
                            <Link
                                id="linkRefreshCompute"
                                onClick={this.onRefreshComputeClick} >
                                <Icon iconName="Refresh" style={{ marginRight: 5, marginLeft: 15 }} />
                                <span>Refresh Compute</span>
                            </Link>
                        }
                    </div>
                </div>

                {!this.props.readOnly &&
                    <div className="form-footer">
                        <DefaultButton text="Cancel" onClick={this.showCancelDialog} />
                        <PrimaryButton type="submit" text="Next" />
                    </div>
                }
            </Form>

            {this.state.showComputeCreator &&
                <ComputeCreator
                    {...this.props}
                    {...this.state}
                    onComputeCreated={this.onComputeCreated}
                    onCancel={this.onComputeCreateCancel}
                />
            }
            <Dialog
                hidden={!this.state.showCancelDialog}
                onDismiss={this.closeCancelDialog}
                dialogContentProps={{
                    type: DialogType.normal,
                    title: "Cancel New Experiment?",
                    subText: `Are you sure you want to cancel creating a new Experiment?`
                }}
                modalProps={{
                    titleAriaId: "cancelDialogLabel",
                    subtitleAriaId: "cancelSubtextLabel",
                    isBlocking: false,
                    containerClassName: "ms-dialogMainOverride"
                }}
            >
                <DialogFooter>
                    <PrimaryButton onClick={this.cancelRun} text="Yes" />
                    <DefaultButton onClick={this.closeCancelDialog} text="No" />
                </DialogFooter>
            </Dialog>
        </>;
    }

    protected readonly getData = async () => {
        this.getExperimentNames();
        this.getComputes();
        this.getVmSizes();
    }

    private readonly onCreateComputeClick = () => {
        this.setState({ showComputeCreator: true });
    }

    private readonly onRefreshComputeClick = () => {
        this.getComputes();
    }

    private readonly onComputeCreated = (compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        this.setState({ showComputeCreator: false, selectedCompute: compute });
        this.getComputes();
    }

    private readonly onComputeCreateCancel = () => {
        this.setState({ showComputeCreator: false });
    }

    private readonly getExperimentNames = async (): Promise<void> => {
        const experiments = await this.services.runHistoryService.listExperiments();
        if (!experiments) {
            return;
        }
        const experimentNames = uniq(map(experiments, (experiment) => experiment.name || ""));
        this.setState({ experimentNames });
    }

    private readonly getComputes = async (): Promise<void> => {
        this.setState({ computeRefreshing: true });
        const computes = await this.services.workSpaceService.listComputes();
        if (!computes) {
            return;
        }
        this.setState({ computes, computeRefreshing: false });
    }

    private readonly getVmSizes = async (): Promise<void> => {
        const vmSizes = await this.services.workSpaceService.listVmSizes();
        if (!vmSizes) {
            return;
        }
        this.setState({ vmSizes });
    }

    private readonly onSubmit = (values: IExperimentStepCallback) => {
        if (!values.computeId || !this.state.computes) {
            return;
        }
        this.props.onNext(values.experimentName, this.state.computes.find((c) => c.id === values.computeId));
    }

    private readonly showCancelDialog = (): void => {
        this.setState({ showCancelDialog: true });
    }

    private readonly closeCancelDialog = (): void => {
        this.setState({ showCancelDialog: false });
    }

    private readonly cancelRun = () => {
        this.props.onCancel();
        this.closeCancelDialog();
    }

}
