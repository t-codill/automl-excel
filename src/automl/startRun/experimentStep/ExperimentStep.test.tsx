import { shallow, ShallowWrapper } from "enzyme";
import { DefaultButton, Dialog, Link, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { Form } from "../../components/Form/Form";
import { RunHistoryService } from "../../services/RunHistoryService";
import { WorkSpaceService } from "../../services/WorkSpaceService";
import { ComputeCreator, IComputeCreatorProps } from "./ComputeCreator";
import { ComputeSelector } from "./ComputeSelector";
import { ExperimentStep, IExperimentStepProps, IExperimentStepState } from "./ExperimentStep";

jest.mock("../../services/WorkSpaceService");
jest.mock("../../services/RunHistoryService");
describe("ExperimentStep", () => {
    let tree: ShallowWrapper<IExperimentStepProps, IExperimentStepState, ExperimentStep>;
    let listComputesSpy: jest.SpyInstance<ReturnType<WorkSpaceService["listComputes"]>>;
    let listVmSizesSpy: jest.SpyInstance<ReturnType<WorkSpaceService["listVmSizes"]>>;
    let listExperimentsSpy: jest.SpyInstance<ReturnType<RunHistoryService["listExperiments"]>>;
    let onNext: jest.Mock;
    let onCancel: jest.Mock;
    beforeEach(async () => {
        onNext = jest.fn();
        onCancel = jest.fn();
        tree = shallow(
            <ExperimentStep
                readOnly={false}
                onEditClick={jest.fn()}
                experimentName={undefined}
                compute={undefined}
                onNext={onNext}
                onCancel={onCancel}
            />);
        await Promise.resolve();
        listComputesSpy = jest.spyOn(WorkSpaceService.prototype, "listComputes");
        listVmSizesSpy = jest.spyOn(WorkSpaceService.prototype, "listVmSizes");
        listExperimentsSpy = jest.spyOn(RunHistoryService.prototype, "listExperiments");
    });
    it("should render", () => {
        expect(tree)
            .toMatchSnapshot();
    });
    it("should pass compute id", () => {
        const root = shallow(
            <ExperimentStep
                readOnly={false}
                onEditClick={jest.fn()}
                experimentName={undefined}
                compute={{
                    id: "computeId"
                }}
                onNext={jest.fn()}
                onCancel={jest.fn()}
            />);
        expect(root.find(ComputeSelector)
            .prop("selectedComputeId"))
            .toBe("computeId");
    });
    it("should show create compute", () => {
        const links = tree.find(Link);
        links
            .filterWhere((l) => {
                return l.prop("id") === "linkCreateCompute";
            })
            .simulate("click");
        expect(tree.state("showComputeCreator"))
            .toBe(true);
    });
    it("should refresh compute", async () => {
        const links = tree.find(Link);
        links
            .filterWhere((l) => {
                return l.prop("id") === "linkRefreshCompute";
            })
            .simulate("click");
        await Promise.resolve();
        expect(listComputesSpy)
            .toBeCalled();
    });

    describe("create compute", () => {
        let propsCreator: IComputeCreatorProps;
        beforeEach(() => {
            tree.setState({ showComputeCreator: true });
            propsCreator = tree.find(ComputeCreator)
                .props();
        });
        it("should show compute create", () => {
            expect(tree.find(ComputeCreator).length)
                .toBe(1);
        });
        it("should call onComputeCreated ", () => {
            const newCompute = {
                id: "newComputeId"
            };
            propsCreator.onComputeCreated(newCompute);
            expect(tree.state())
                .toEqual(expect.objectContaining({
                    showComputeCreator: false,
                    selectedCompute: newCompute
                }));
        });
        it("should call onComputeCreateCancel ", () => {
            propsCreator.onCancel();
            expect(tree.state())
                .toEqual(expect.objectContaining({
                    showComputeCreator: false
                }));
        });
    });

    it("should get experiment name", async () => {
        expect(tree.state("experimentNames"))
            .toEqual(["Experiment Name 1", "Experiment Name 2", ""]);
    });

    it("should still refreshing experiment name", async () => {
        listExperimentsSpy.mockReturnValueOnce(Promise.resolve(undefined));
        const refreshingTree = shallow<ExperimentStep>(
            <ExperimentStep
                readOnly={false}
                onEditClick={jest.fn()}
                experimentName={undefined}
                compute={undefined}
                onNext={jest.fn()}
                onCancel={jest.fn()}
            />);
        await Promise.resolve();
        expect(refreshingTree.state("experimentNames"))
            .toBeUndefined();
    });

    it("should get computes", async () => {
        expect(tree.state())
            .toEqual(expect.objectContaining({
                computeRefreshing: false,
                computes: [{
                    id: "Test_ID",
                    location: "eastus",
                    name: "Test_Compute",
                }]
            }));
    });

    it("should still refreshing computes", async () => {
        listComputesSpy.mockReturnValueOnce(Promise.resolve(undefined));
        const refreshingTree = shallow<ExperimentStep>(
            <ExperimentStep
                readOnly={false}
                onEditClick={jest.fn()}
                experimentName={undefined}
                compute={undefined}
                onNext={jest.fn()}
                onCancel={jest.fn()}
            />);
        await Promise.resolve();
        expect(refreshingTree.state("computes"))
            .toBeUndefined();
        expect(refreshingTree.state("computeRefreshing"))
            .toBe(true);
    });

    it("should get vm sizes", async () => {
        expect(tree.state("vmSizes"))
            .toEqual({
                0: {
                    family: "standardDSv2Family",
                    lowPriorityCapable: true,
                    maxResourceVolumeMB: 7168,
                    memoryGB: 3.5,
                    name: "Standard_DS1_v2",
                    osVhdSizeMB: 1047552,
                    premiumIO: true,
                    vCPUs: 1,
                },
                amlCompute: [
                    {
                        family: "standardDSv2Family",
                        lowPriorityCapable: true,
                        maxResourceVolumeMB: 7168,
                        memoryGB: 3.5,
                        name: "Standard_DS1_v2",
                        osVhdSizeMB: 1047552,
                        premiumIO: true,
                        vCPUs: 1,
                    },
                ],
            });
    });

    it("should still refreshing vm sizes", async () => {
        listVmSizesSpy.mockReturnValueOnce(Promise.resolve(undefined));
        const refreshingTree = shallow<ExperimentStep>(
            <ExperimentStep
                readOnly={false}
                onEditClick={jest.fn()}
                experimentName={undefined}
                compute={undefined}
                onNext={jest.fn()}
                onCancel={jest.fn()}
            />);
        await Promise.resolve();
        expect(refreshingTree.state("vmSizes"))
            .toBeUndefined();
    });

    it("should submit", async () => {
        const value = {
            computeId: "Test_ID",
            experimentName: "testName"
        };
        const onSubmit = tree.find(Form)
            .prop("onSubmit");
        onSubmit(value);
        expect(onNext)
            .toBeCalledWith("testName", expect.objectContaining({ id: "Test_ID" }));
    });

    it("should not submit without compute id", async () => {
        const value = {
            computeId: undefined,
            experimentName: "testName"
        };
        const onSubmit = tree.find(Form)
            .prop("onSubmit");
        onSubmit(value);
        expect(onNext)
            .not
            .toBeCalled();
    });

    describe("cancel", () => {
        beforeEach(() => {
            tree.find(DefaultButton)
                .filterWhere((b) => b.prop("text") === "Cancel")
                .simulate("click");
        });
        it("should show cancel dialog", () => {
            expect(tree.state("showCancelDialog"))
                .toBe(true);
            expect(tree.find(Dialog).length)
                .toBe(1);
        });
        it("should cancel when click yes", () => {
            tree.find(PrimaryButton)
                .filterWhere((b) => b.prop("text") === "Yes")
                .simulate("click");
            expect(onCancel)
                .toBeCalled();
            expect(tree.state("showCancelDialog"))
                .toBe(false);
        });
        it("should not cancel when click no", () => {
            tree.find(DefaultButton)
                .filterWhere((b) => b.prop("text") === "No")
                .simulate("click");
            expect(onCancel)
                .not
                .toBeCalled();
            expect(tree.state("showCancelDialog"))
                .toBe(false);
        });
    });
});
