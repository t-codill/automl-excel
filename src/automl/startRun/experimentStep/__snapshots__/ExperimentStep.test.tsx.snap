// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentStep should render 1`] = `
<Fragment>
  <Form
    onSubmit={[Function]}
  >
    <div
      className="start-run-step-content"
    >
      <div
        className="ms-md12 ms-lg8 ms-xl6 ms-xxl4"
      >
        <ExperimentNameInput
          experimentNames={
            Array [
              "Experiment Name 1",
              "Experiment Name 2",
              "",
            ]
          }
          onEditClick={[MockFunction]}
          readonly={false}
        />
      </div>
      <div
        className="ms-md12 ms-lg8 ms-xl6 ms-xxl4"
      >
        <ComputeSelector
          computes={
            Array [
              Object {
                "id": "Test_ID",
                "location": "eastus",
                "name": "Test_Compute",
              },
            ]
          }
          onEditClick={[MockFunction]}
          readonly={false}
          refreshing={false}
        />
        <StyledLinkBase
          id="linkCreateCompute"
          onClick={[Function]}
        >
          <StyledIconBase
            iconName="Devices3"
            style={
              Object {
                "marginRight": 5,
              }
            }
          />
          <span>
            Create a new compute
          </span>
        </StyledLinkBase>
        <StyledLinkBase
          id="linkRefreshCompute"
          onClick={[Function]}
        >
          <StyledIconBase
            iconName="Refresh"
            style={
              Object {
                "marginLeft": 15,
                "marginRight": 5,
              }
            }
          />
          <span>
            Refresh Compute
          </span>
        </StyledLinkBase>
      </div>
    </div>
    <div
      className="form-footer"
    >
      <CustomizedDefaultButton
        onClick={[Function]}
        text="Cancel"
      />
      <CustomizedPrimaryButton
        text="Next"
        type="submit"
      />
    </div>
  </Form>
  <StyledWithResponsiveMode
    dialogContentProps={
      Object {
        "subText": "Are you sure you want to cancel creating a new Experiment?",
        "title": "Cancel New Experiment?",
        "type": 0,
      }
    }
    hidden={true}
    modalProps={
      Object {
        "containerClassName": "ms-dialogMainOverride",
        "isBlocking": false,
        "subtitleAriaId": "cancelSubtextLabel",
        "titleAriaId": "cancelDialogLabel",
      }
    }
    onDismiss={[Function]}
  >
    <StyledDialogFooterBase>
      <CustomizedPrimaryButton
        onClick={[Function]}
        text="Yes"
      />
      <CustomizedDefaultButton
        onClick={[Function]}
        text="No"
      />
    </StyledDialogFooterBase>
  </StyledWithResponsiveMode>
</Fragment>
`;
